proposals/NNNN-codable-error-printing.md

```diff
@@ -0,0 +1,160 @@
+
+## Proposed solution
+
+Conform `EncodingError` and `DecodingError` to `CustomDebugStringConvertible` and provide a clean, readable debug description for each. Here is an example of the proposed change for the same decoding error as above.
+
+```
+Key 'population' not found in keyed decoding container.
+Debug description: No value associated with key CodingKeys(stringValue: "population", intValue: nil) ("population").
+Path: [0]/home/country
+```
+
+Complete examples of the before/after diffs are available in the description of the [implementation pull request](https://github.com/swiftlang/swift/pull/80941) that accompanies this proposal.
+
+**Note 1:** this proposal is _not_ intended to specify an exact output format. The above is provided as an example, and is not a guarantee of current or future behavior. You are still free to inspect the contents of thrown errors directly if you need to detect specific problems.
+
+**Note 2:** the output could be further improved by modifying `JSONDecoder` to write a better debug description. See [Future Directions](#future-directions) for more.
+
+## Detailed design
+
+```swift
+@available(SwiftStdlib 6.2, *)
+extension EncodingError: CustomDebugStringConvertible {
+  public var debugDescription: String {...}
+}
+
+@available(SwiftStdlib 6.2, *)
+extension DecodingError: CustomDebugStringConvertible {
+  public var debugDescription: String {...}
+}
+```
+
+## Source compatibility
+
+The new conformance changes the result of converting an `EncodingError` or `DecodingError` value to a string. This changes observable behavior: code that attempts to parse the result of `String(describing:)` or `String(reflecting:)` can be misled by the change of format.
+
+However, the documentation of these interfaces explicitly state that when the input type conforms to none of the standard string conversion protocols, then the result of these operations is unspecified.
+
+Changing the value of an unspecified result is not considered to be a source incompatible change.
+
+## ABI compatibility
+
+The proposal conforms two previously existing stdlib types to a previously existing stdlib protocol. This is technically an ABI breaking change: on ABI-stable platforms, we may have preexisting Swift binaries that implement a retroactive `CustomDebugStringConvertible` conformance, or binaries that assume that the existing error types do _not_ conform to the protocol.
+
+We do not expect this to be an issue in practice. The above checks represent edge cases, since checking an arbitrary error for conformance to `CustomDebugStringConvertible` at run-time seems unlikely. And if someone has added a conformance, when they re-build with a stdlib that includes this conformance, they will get a build-time warning.
```

Comment:
```
And if someone has added a conformance, when they re-build with a stdlib that includes this conformance, they will get a build-time warning.
```

What happens when you recompile doesn't make ABI breakage on ABI-stable platforms any more or less problematic, so I don't think it needs comment here...

Reply:
I felt weird about just stating that we don't expect it to be an issue but [not explaining why](https://xkcd.com/979/), but I've tightened up the wording. I figure, this section will be most relevant if someone in the future _is_ negatively affected by the change, and in that case, I don't want them finding a years-old document telling them that their problem is unlikely 😆